pegen/grammar_visualizer.py

```diff
@@ -0,0 +1,92 @@
+import argparse
+import tokenize
+import enum
+
+from pegen.tokenizer import Tokenizer
+from pegen.tokenizer import grammar_tokenizer
+from pegen.grammar2 import GrammarParser
+
+
+argparser = argparse.ArgumentParser(prog='pegen', description="Experimental PEG-like parser generator")
+argparser.add_argument('filename', help="Grammar description")
+
+
+class Direction(enum.Enum):
+        UP = 0
+        DOWN = 1
+        NEUTRAL = 2
+        NONE = 3
+
+
+class GrammarPrinter:
+
+    def children(self, node):
+        for value in node:
+            if isinstance(value, list):
+                yield from value
+            else:
+                yield value
+    def nb_children(self, node):
+        return sum(self.nb_children(child) for child in self.children(node)) + 1
+
+    def name(self, node):
+        if (name := getattr(node, "name", None)):
+            return node.__class__.__name__ + f"({node.name})"
+        elif (value := getattr(node, "value", None)):
+            return node.__class__.__name__ + f"({node.value})"
+        else:
+            return node.__class__.__name__
+
+
+    def print_grammar(self, current_node, indent='', last=Direction.NEUTRAL, printer=print):
```

Comment:
This method has too much whitespace in it (to my taste :-).

Do you have a reference for the algorithm?

Reply:
I have done a refactor of everything, hopefully, is now a bit more clear. If not I can explain or add more comments.